src/NATS.Client.Core/NatsBufferWriter.cs

```diff
@@ -0,0 +1,451 @@
+    public NatsBufferWriter(int initialCapacity)
+        : this(ArrayPool<T>.Shared, initialCapacity)
+    {
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="NatsBufferWriter{T}"/> class.
+    /// </summary>
+    /// <param name="pool">The <see cref="ArrayPool{T}"/> instance to use.</param>
+    /// <param name="initialCapacity">The minimum capacity with which to initialize the underlying buffer.</param>
+    /// <exception cref="ArgumentOutOfRangeException">Thrown when <paramref name="initialCapacity"/> is not valid.</exception>
+    public NatsBufferWriter(ArrayPool<T> pool, int initialCapacity)
+    {
+        // Since we're using pooled arrays, we can rent the buffer with the
+        // default size immediately, we don't need to use lazy initialization
+        // to save unnecessary memory allocations in this case.
+        // Additionally, we don't need to manually throw the exception if
+        // the requested size is not valid, as that'll be thrown automatically
+        // by the array pool in use when we try to rent an array with that size.
+        _pool = pool;
+        _array = pool.Rent(initialCapacity);
+        _index = 0;
+    }
+
+    /// <inheritdoc/>
+    Memory<T> IMemoryOwner<T>.Memory
+    {
+        // This property is explicitly implemented so that it's hidden
+        // under normal usage, as the name could be confusing when
+        // displayed besides WrittenMemory and GetMemory().
+        // The IMemoryOwner<T> interface is implemented primarily
+        // so that the AsStream() extension can be used on this type,
+        // allowing users to first create a NatsBufferWriter<byte>
+        // instance to write data to, then get a stream through the
+        // extension and let it take care of returning the underlying
+        // buffer to the shared pool when it's no longer necessary.
+        // Inlining is not needed here since this will always be a callvirt.
+        get => MemoryMarshal.AsMemory(WrittenMemory);
+    }
+
+    /// <inheritdoc/>
+    public ReadOnlyMemory<T> WrittenMemory
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        get
+        {
+            var array = _array;
+
+            if (array is null)
+            {
+                ThrowObjectDisposedException();
+            }
+
+            return array!.AsMemory(0, _index);
+        }
+    }
+
+    /// <inheritdoc/>
+    public ReadOnlySpan<T> WrittenSpan
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        get
+        {
+            var array = _array;
+
+            if (array is null)
+            {
+                ThrowObjectDisposedException();
+            }
+
+            return array!.AsSpan(0, _index);
+        }
+    }
+
+    /// <inheritdoc/>
+    public int WrittenCount
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
Are those inlining hints necessary? Presumably JIT should always inline simple field accesses and the other getters look like they are likely inlined without `AggressiveInlining` when on a hot path.
I think it would be better to remove the inlining hint here and from all other getters, unless there is evidence that these aren't inined when on a hot path, and they are inlined when marked with `AggressiveInlining`, and there is a measurable performance gain by doing so.
What would help with performance is to guarantee that JIT always knows the concrete target type by sealing this type and removing the interface.

Reply:
done